### PR TITLE
Lint single-character strings as P: Pattern args

### LINKED ISSUE #650: lint .split("_") on constant single-char string
Suggest `_.split('x')` instead. Should be easy enough with const_eval.

### PR DESCRIPTION
**Pull request by shssoichiro**
Fixes #650

### CODE REVIEW

**Comment by shssoichiro** [comment]
I'd like to go back and get this working with multi-byte characters as well, but it's almost impossible to debug ICE (see comments in test suite) on OS X because the backtraces don't have filenames or line numbers.

File: src/methods.rs [hunk 1]
```diff
@@ -817,6 +839,22 @@ fn lint_chars_next(cx: &LateContext, expr: &Expr, chain: &Expr, other: &Expr, eq
...
 }
 
+/// lint for length-1 `str`s for methods in `PATTERN_METHODS`
+fn lint_single_char_pattern(cx: &LateContext, expr: &Expr, arg: &Expr) {
+    if let Ok(ConstVal::Str(r)) = eval_const_expr_partial(cx.tcx, arg, ExprTypeChecked, None) {
+        if r.len() == 1 {
+            let hint = snippet(cx, expr.span, "..").replace(&format!("\"{}\"", r), &format!("'{}'", r));
+            span_lint_and_then(cx,
+                               SINGLE_CHAR_PATTERN,
+                               expr.span,
```
Comment (mcarton) [inline src/methods.rs]: That one should be `arg.span`.

File: src/methods.rs [hunk 2]
```diff
@@ -295,6 +297,20 @@ declare_lint! {
     pub NEW_RET_NO_SELF, Warn, "not returning `Self` in a `new` method"
 }
 
+/// **What it does:** This lint checks for string methods that receive a single-character `str` as an argument, e.g. `_.split("x")`
+///
+/// **Why is this bad?** Performing these methods using a `str` may be slower than using a `char`
+///
+/// **Known problems:** Does not catch multi-byte unicode characters
```
Comment (mcarton) [inline src/methods.rs]: Small nit: other wiki entries ends with a dot.

**Comment by shssoichiro** [comment]
New commit pushed with @mcarton's nits fixed.

File: tests/compile-fail/methods.rs [hunk 3]
```diff
@@ -355,3 +355,92 @@ fn clone_on_double_ref() {
...
+    x.split("xx");
+
+    x.split('x');
+
+    let y = "x";
+    x.split(y);
+
+    // Not yet testing for multi-byte characters
+    // Changing `r.len() == 1` to `r.chars().count() == 1` in `lint_single_char_pattern`
+    // should have done this but produced an ICE
```
Comment (llogiq) [inline tests/compile-fail/methods.rs]: Is there a bug open for that ICE?

File: tests/compile-fail/methods.rs [hunk 4]
```diff
@@ -355,3 +355,92 @@ fn clone_on_double_ref() {
...
+    //~| SUGGESTION x.rmatch_indices('x');
+    x.trim_left_matches("x");
+    //~^ ERROR single-character string constant used as pattern
+    //~| HELP try using a char instead:
+    //~| SUGGESTION x.trim_left_matches('x');
+    x.trim_right_matches("x");
+    //~^ ERROR single-character string constant used as pattern
+    //~| HELP try using a char instead:
+    //~| SUGGESTION x.trim_right_matches('x');
+}
```
Comment (llogiq) [inline tests/compile-fail/methods.rs]: Small nit: It is customary to end files with a newline.

File: tests/compile-fail/methods.rs (hunk 3 above)
Comment (shssoichiro) [inline tests/compile-fail/methods.rs]: I couldn't find one, although there may be (there are a lot of issues on rustc to filter through). It seems to happen here:

    tests/compile-fail/methods.rs:377:13: 377:16 help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#single_char_pattern
    tests/compile-fail/methods.rs:378:13: 378:16 error: single-character string constant used as pattern
    tests/compile-fail/methods.rs:378     x.split("💣");
                                                  ^~~
    tests/compile-fail/methods.rs:4:9: 4:15 note: lint level defined here
    tests/compile-fail/methods.rs:4 #![deny(clippy, clippy_pedantic)]
                                            ^~~~~~
    tests/compile-fail/methods.rs:378:5: 378:17 help: try using a char instead:
    error: internal compiler error: Error constructed but not emitted
    thread panicked while panicking. aborting.

Probably related, the formatting on the hint output gets broken on multibyte chars, e.g.:

    tests/compile-fail/methods.rs:377:13: 377:16 error: single-character string constant used as pattern
    tests/compile-fail/methods.rs:377     x.split("ℝ");
                                                  ^~~
    tests/compile-fail/methods.rs:4:9: 4:15 note: lint level defined here
    tests/compile-fail/methods.rs:4 #![deny(clippy, clippy_pedantic)]
                                            ^~~~~~
    tests/compile-fail/methods.rs:377:5: 377:17 help: try using a char instead:
    tests/compile-fail/methods.rs:        x.split('ℝ')");

**Pull request merged**